zenml/steps/trainer/utils.py

```diff
@@ -64,7 +64,7 @@ def combine_batch_results(x):
     for batch in x:
         for feature, values in batch.items():
             if isinstance(values, torch.Tensor):
-                temp = torch.squeeze(values).detach().numpy()
+                temp = torch.squeeze(values).detach().cpu().numpy()
```

Comment:
I'm not sure super sure whether we want to make this explicitly CPU here. Perhaps adding a flag in the trainer itself would make sense?

Reply:
> Thank you very much for the contribution! I like the idea but perhaps we can make it more an explicitly defined behavior in the trainer constructor rather than an explicit behavior. That way people with GPUs and CPUs both could use it. What do you think?

Agree completely. I am thinking of keeping `device` as an attribute of `FeedForwardTrainer` whose default value should be `torch.device("cpu")` and users have the choice to pass the value of this argument also while initializing trainer. `TorchBaseTrainerStep` has no implementation otherwise I was thinking of keeping `device` attribute there since for every `PyTorch` model `device` needs to be configured.

Reply:
Makes sense. As a bonus: Can we infer from the environment whether a GPU is available and then set the default value accordingly?

Reply:
> I'm not sure super sure whether we want to make this explicitly CPU here. Perhaps adding a flag in the trainer itself would make sense?

If a `tensor` is already on the desired device doing `.cpu()` or `.cuda()` has no effect in `PyTorch` so this would work fine without the need of any `device` flag.

Reply:
Yes but if someone does have a GPU and wants to use it then they should be able to default to it, no?

Reply:
> Makes sense. As a bonus: Can we infer from the environment whether a GPU is available and then set the default value accordingly?

Yes, we sure we can. Will do.

Reply:
> Yes but if someone does have a GPU and wants to use it then they should be able to default to it, no?

I need a bit more context here @htahir1. So `temp` variable is going to be used to create `results` `dict` object which contains the output of `model` on `test_dataset`. This `result` object is being saved as `TFRecords` file in the current code. I am assuming (not much familiar with `tfma`) `result` object or its equivalent `TFRecods` files will be used in evaluation step. If that is so putting `results` on GPU will just consume VRAM without any computation being performed on them. Is there any other reason to use GPU for this? 

Also, `to_serialized_examples` function itself needs to be changed or another function needs to be added to convert `torch.tensors`  to `floats, ints` or `bytes` before calling `to_serialized_examples`.

Reply:
@SKRohit Sorry for the late response. I think @bcdurak knows this part of the code better. @bcdurak would you say we should take this to GPU or stick to CPU?

Reply:
@htahir1 and @bcdurak I have made `device` a `trainer` attribute. If you will let me know I would do the changes in the code to handle `result` in GPU itself.

Reply:
`to_serialized_examples` works because we convert to numpy and it can handle that. Your changes now look great except one small detail: Currently ZenML Steps can only accept primitives as input (ints, floats, strings etc) so a `torch.device` won't work. Can we replace the parameter with a boolean (i.e. `gpu` true or false) or a string like `gpu` / `cpu:0`? Otherwise looks great!

Reply:
Yes. But if we use GPU instead of numpy arrays we would have PyTorch tensors and then we would need to change `to_serialize_examples`. So better keep to handle `result` on CPU itself. 

Btw I have implemented what you suggested to change. I have made `device` to be an `str`. I was testing the code and found that somehow the value of `device` provided by user is not used. `self.steps_dict['trainer']['args']` are still being populated with 

`'{"split_mapping": {"train_splits": ["train"], "eval_splits": ["eval"], "test_splits": ["test"]}, "batch_size": 32, "lr": 0.0001, "epochs": 100, "dropout_chance": 0.2, "loss": "binary_crossentropy", "metrics": ["accuracy"], "hidden_layers": [64, 32, 16], "hidden_activation": "relu", "last_activation": "sigmoid", "input_units": 8, "output_units": 1}'` There is no value of `device` even if I passed while adding trainer   `training_pipeline.add_trainer(TorchFeedForwardTrainer(
    loss='binary_crossentropy',
    last_activation='sigmoid',
    output_units=1,
    metrics=['accuracy'],
    device="cuda:0",
    epochs=100))`. 

Could you please let me know why this is happening and where should I look to understand how `self.steps_dict` is being populated?

Reply:
I think its because you need to pass the variable to the `super()` function e.g. `super(device=device...)` at the end of the constructor of your step.

Reply:
Yes, passing `device` to `super` worked. I have done the changes. Let me know if it is fine.